src/enzo/ActiveParticleRoutines.C

```diff
@@ -0,0 +1,508 @@
+  dx = LengthUnits * CurrentGrid->CellWidth[0][0];
+  MassConversion = (float) (dx*dx*dx * double(DensityUnits) / Msun);
+  this->Mass *= MassConversion;
+  return;
+}
+
+void  ActiveParticleType::AdjustVelocity(float VelocityIncrement[])
+{ 
+  int i;
+  for (i = 0; i<3; i++)
+    vel[i] += VelocityIncrement[i];
+  return;
+}
+
+void ActiveParticleType::SetVelocity(float NewVelocity[])
+{
+  int i;
+  for (i = 0; i<3; i++)
+    vel[i] = NewVelocity[i];
+  return;
+}
+
+void ActiveParticleType::SetPosition(FLOAT NewPosition[])
+{
+  int i;
+  for (i = 0; i<3; i++)
+    pos[i] = NewPosition[i];
+  return;
+}
+
+void ActiveParticleType::SetPositionPeriod(FLOAT period[])
+{
+  int i;
+  for (i = 0; i<3; i++) {
+    pos[i] = fmod(pos[i], period[i]);
+    if (pos[i] < 0) {
+      pos[i] += period[i];
+    }
+  }
+  return;
+}
+
+void ActiveParticleType::Merge(ActiveParticleType *a)
+{
+  int dim;
+  double ratio1, ratio2;
+  ratio1 = Mass / (Mass + a->Mass);
+  ratio2 = 1.0 - ratio1;
+  Metallicity = ratio1 * Metallicity + ratio2 * a->Metallicity;
+  for (dim = 0; dim < MAX_DIMENSION; dim++) {
+    pos[dim] = ratio1 * pos[dim] + ratio2 * a->pos[dim];
+    vel[dim] = ratio1 * vel[dim] + ratio2 * a->vel[dim];
+    //accreted_angmom[dim] = ratio1 * accreted_angmom[dim] + ratio2 * a.accreted_angmom[dim];
+  }
+  Mass += a->Mass;
+  //FinalMass += a.FinalMass;
+  //DeltaMass += a.DeltaMass;
+  //last_accretion_rate += a.last_accretion_rate;
+  //NotEjectedMass += a.NotEjectedMass;
+  return;
+}
+
+
+/*
+ * This is a little bit more advanced merging. 
+ * If all depends on which star retains the original 
+ * attributes.
+ * This star becomes the "new star" if 
+ * 1. It's mass is greater than the "a" star.
+ */
+void ActiveParticleType_SmartStar::SmartMerge(ActiveParticleType_SmartStar *a)
+{
+  int dim;
+  double ratio1, ratio2;
+
+  ratio1 = Mass / (Mass + a->Mass);
+  ratio2 = 1.0 - ratio1;
+  Metallicity = ratio1 * Metallicity + ratio2 * a->Metallicity;
+  for (dim = 0; dim < MAX_DIMENSION; dim++) {
+    pos[dim] = ratio1 * pos[dim] + ratio2 * a->pos[dim];
+    vel[dim] = ratio1 * vel[dim] + ratio2 * a->vel[dim];
+ 
+  }
+  if(Mass > a->Mass) {
+    ;
+  }
+  else {
+    BirthTime = a->BirthTime;
+    oldmass = a->oldmass;
+    TimeIndex = a->TimeIndex;
+    for(int i = 0; i < NTIMES; i++)
+      {
+	AccretionRateTime[i] = a->AccretionRateTime[i];
+	AccretionRate[i] = a->AccretionRate[i];	
+      }
+  }
+  Mass += a->Mass;
+  NotEjectedMass += a->NotEjectedMass;
+  ParticleClass = max(ParticleClass, a->ParticleClass);
+  return;
+}
+
+bool ActiveParticleType::Mergable(ActiveParticleType *a)
+{
+  // Only merge yet-to-be born stars
+  return type == a->type && type < 0;
+}
+
+#ifdef UNUSED
+bool ActiveParticleType::MergableMBH(ActiveParticleType *a)
+{
+  // Merge MBH particle with another 
+  return type == a->type && type == MBH;
+}
+#endif
+
+float ActiveParticleType::Separation2(ActiveParticleType *a)
+{
```

Comment:
Add a comment about the purpose of this function.

Reply:
Done